### PR TITLE
feat: cache assets with service worker

### PR DESCRIPTION
**Pull request by patriciagarcia**

### CODE REVIEW

**Comment by gr2m** [comment]
I’m amazed how simple & powerful this is.

I think we should add the `sw.js` code to every HTML file though, right?

File: public/sw.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    caches.match(event.request).then(function(response) {
+      return response || fetch(event.request);
+    })
+  );
+});
+
+// 'activate' is fired when the service worker is ready to handle
+// events.
+self.addEventListener('activate', function(event) {
+  console.log('active! now you can work offline');
```
Comment (NickColley) [inline public/sw.js]: Should we tell the user they can work offline or just remove this console.log?

**Comment by NickColley** [comment]
Almost wonder if it's better to just use appcache here (don't shoot me), appcache works when you're doing this kind of very simple approach.

Will work better cross browser too.

File: public/sw.js (hunk 1 above)
Comment (patriciagarcia) [inline public/sw.js]: @nickcolley yes! we wanted to tell the users they were offline, but we haven't yet decided how to do it, so I just left this `console.log` to show where to do it. I can remove it before merging.

**Comment by patriciagarcia** [comment]
@nickcolley you have a point there.. the only thing is that, since we're trying to show them how to do it themselves, I think it's better to show them service worker than a deprecated technology.

It's actually very funny, if you go to mdn, when looking at appcache [you get a warning](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache) because it's deprecated and a suggestion to use service worker, but when you go to service worker [you get a warning](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) also because it's still experimental :grin: . Not sure what they expect you to do with it!

**Comment by patriciagarcia** [comment]
@gr2m not the service worker `sw.js` itself but maybe the registration code? (in `index.html`). If they start on the first page, it doesn't matter, since the service worker is added and activated and then all the other `.html` files are cached. But if they start in another page, that's when there can be a problem.

Honestly I am not sure what's the pattern here but there is probably no harm in adding it everywhere.

**Comment by patriciagarcia** [comment]
I pushed a version with the service worker registered everywhere, if you like it I can squash the commits.

**Comment by gr2m** [comment]
I added two steps for service worker and connection status. I also reverted adding the service worker code to every page, just having it on the first page has the advantage that we can embed the code and explain it. Sorry for going back and forth :( 

Let me know how you like the new first & last pages.

I’m also wondering: right now, after the install, service worker does not update itself. Can we change that somehow? Is there a way that service worker can update the cached assets, e.g. on every page load? Don’t know if it makes sense. At least when we change the `v1` to something else in `sw.js` it should update the cache I’d say

**Comment by gr2m** [comment]
What I would like to add is a page in the end where tell user to sign in to their account in two different browsers, then start adding colors – they will sync from one to another. Now turn of wifi, continue adding colors, no sync for obvious reasons. Now turn on wifi again ... wait for it ... MAGIC :star2:

**Comment by patriciagarcia** [comment]
The service worker gets updated every time that `sw.js` changes, however, when you do update it things gets a bit more messy, since pages that were already open stay with the old service browser, [while new pages go the new one](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers). So yes, in a real app the `sw.js` code won't be so simple, but I think for an introduction simple is better. I think it's in a good level with the kind of introduction we're doing for Hoodie, in a real app you'll be doing much more complicated things with Hoodie itself, but we just want to give a feel of what is possible. 

I'd personally leave it as it is, or if not, I'd just do something simple like [deleting all caches when installing a new service worker](https://github.com/jakearchibald/simple-serviceworker-tutorial/blob/gh-pages/reset/sw.js), before opening a new one.

> I’m also wondering: right now, after the install, service worker does not update itself. Can we change that somehow? Is there a way that service worker can update the cached assets, e.g. on every page load? Don’t know if it makes sense. At least when we change the v1 to something else in sw.js it should update the cache I’d say

Also, tbh I am not quite sure yet what's the proper way of detecting changes in cached files with service workers (I just started learning it myself!). Should we check on every request or on every 'install' or 'activate' event if the assets have changed? should we invalidate whole caches? (doesn't seem too efficiente? or should we add a unique id to assets that we can update every time assets change? (like `3.html-v2`)?

File: public/index.html [hunk 2]
```diff
@@ -29,32 +29,49 @@
...
 
 <script>
-$('#formColor').submit(function(event) {
-  var color = $('#formColor input').val();
-  $(document.body).css('background', color);
+if ('serviceWorker' in navigator) {
+  // Register a service worker to allow the tutorial to work offline
+  navigator.serviceWorker.register('/sw.js', {
+    // The scope cannot be parent to the script url
+    scope: '../../'
```
Comment (patriciagarcia) [inline public/index.html]: `scope: './'`, as said in the code `The scope cannot be parent to the script ulr` ;-).

File: public/index.html [hunk 3]
```diff
@@ -29,32 +29,49 @@
...
+  })
+
+  .then(function () {
+    $('.service-worker-status').text('ready!')
+  })
```
Comment (patriciagarcia) [inline public/index.html]: mm, I am not sure this works like that, when does `register` promise return? I think the service worker is "ready" when it's `activated`, not when it's [registered](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers). I think the text about the status should be updated [here](https://github.com/hoodiehq/hoodie-camp-tutorial/pull/3/files#diff-7a7a9fab4ca3224cad19254b6aba0be3R55), or if you want to be more precise about the state, then we can use [the `statechange` event](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/onstatechange).

File: public/sw.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  'assets/img/pattern_130.gif',
+  'assets/js/main.js',
+  'assets/vendor/bootstrap/bootstrap.js',
+  'assets/vendor/bootstrap/bootstrap.min.css',
+  'assets/vendor/prism/prism.js',
+  'assets/vendor/prism/prism.css',
+  'assets/vendor/bootstrap.modalform.js',
+  'assets/vendor/jquery-2.1.0.min.js',
+  'hoodie/client.js',
+  'index.html',
```
Comment (patriciagarcia) [inline public/sw.js]: if the service worker is always added in `index.html` there is no need to cache that (that will happen automatically). I think there is a mix of code here between my first and second commit, you can just forgot about my second commit altogether if we're just going to add it in the first page ;-).

**Comment by patriciagarcia** [comment]
Pushed a version with a couple of the problems fixed, that empties obsolete cache versions on ServiceWorker installation.

Btw, great idea! :-)

> What I would like to add is a page in the end where tell user to sign in to their account in two different browsers, then start adding colors – they will sync from one to another. Now turn of wifi, continue adding colors, no sync for obvious reasons. Now turn on wifi again ... wait for it ... MAGIC :star2:

File: public/index.html (hunk 3 above)
Comment (gr2m) [inline public/index.html]: can we access the DOM from https://github.com/hoodiehq/hoodie-camp-tutorial/pull/3/files#diff-7a7a9fab4ca3224cad19254b6aba0be3R55?

**Comment by gr2m** [comment]
sorry didn’t get to add the last page yet, will do on Monday

**Comment by patriciagarcia** [comment]
More updates! I made the service worker code nicer and easier to understand, with comments and console.logs, in case they want to check what's happening. Do you think it's too much? 

Also, I fixed the problem of updating the service worker status in the proper way.

**Comment by patriciagarcia** [comment]
One thing, I can't get the code in page 2 to show anything else than "online" all the time. Is it working for you?

Removing all colors in page 6 is also  not removing the background color, although I am not sure if that is intentioned.

**Comment by patriciagarcia** [comment]
Btw, this is the messiest PR I've ever seen :D.

I tested everything (online, offline) but it would be nice if somebody else can give it another try, just in case I missed something.

Going back to slides now!

**Comment by gr2m** [comment]
changes looking good :) Don’t worry we will clean up the PR at the end. Testing now

**Comment by gr2m** [comment]
added a few minor improvements. On to create final page for sync magic ✨

**Pull request merged**